q2doc/usage/usage.py

```diff
@@ -0,0 +1,200 @@
+        factories_to_nodes(block, env)
+    for record in records:
+        if record.source not in ["init_metadata", "init_data"]:
+            continue
+        ext = get_ext(record.source)
+        artifact = record.result
+        out_dir = Path(env.app.outdir) / node.docname / 'results'
+        if not out_dir.exists():
+            out_dir.mkdir(parents=True)
+        fp = out_dir / f'{record.ref}.{ext}'
+        artifact.save(str(fp))
+
+
+@records_to_nodes.register(CLIUsage)
+def cli(use, records, block, env):
+    for record in records:
+        if record.source == "action":
+            rendered = "\n\n".join(use.render())
+            example = remove_rendered(rendered, env.rendered['cli'])
+            env.rendered['cli'] = rendered
+            node = UsageExampleNode(cli=example)
+            block["nodes"] = [node]
+            # Break after seeing the first record created by use.action() since
+            # we only need to call use.render() once.
+            break
+
+
+@records_to_nodes.register(ArtifactAPIUsage)
+def artifact_api(use, records, block, env):
+    for record in records:
+        source = record.source
+        if source == "init_data":
+            data_node = init_data_node(record, env)
+            block['nodes'].append(data_node)
+        elif source == "init_metadata":
+            metadata_node = init_data_node(record, env)
+            block['nodes'].append(metadata_node)
+        elif source == "action":
+            node = block["nodes"][0]
+            setup_code = get_data_nodes(env)
+            rendered = use.render()
+            example = remove_rendered(rendered, env.rendered['art_api'])
+            example = setup_code + example
+            env.rendered['art_api'] = rendered
+            node.artifact_api = example
+            # Break after seeing the first record created by use.action() since
+            # we only need to call use.render() once.
+            break
+
+
+def init_data_node(record, env):
+    name = record.ref
+    ext = get_ext(record.source)
+    fname = f"{name}.{ext}"
+    exc_use = env.domaindata['q2']['drivers']['exc_use']
+    result = exc_use._get_record(name).result
+    type_ = "Artifact" if isinstance(result, qiime2.Artifact) else "Metadata"
+    load_statement = f"{name} = {type_}.load('{fname}')"
+    node = UsageDataNode(load_statement, name=name)
+    return node
+
+
+def get_data_nodes(env):
+    setup_code = []
+    for block in env.usage_blocks:
+        nodes = block['nodes']
+        for node in nodes:
+            if isinstance(node, UsageDataNode) and not node.loaded:
+                setup_code.append(node.setup)
+                node.loaded = True
+    return '\n'.join(sorted(setup_code)) + '\n\n'
+
+
+def remove_rendered(example, rendered):
+    imports = [line for line in rendered.splitlines() if 'import' in line]
+    # TODO Don't remove previously rendered variable binding in case the same
+    #   name is overwritten
```

Comment:
e.g., if somebody writes `   ints_a = use.init_data('ints_a', ints1_factory)` for one example, then  reuses `ints_a` for another example, but gives it another value like `ints_a = use.init_data('ints_b', ints2_factory)`.